src/operators.jl

```diff
@@ -83,120 +127,152 @@ function Base.:^(lhs::Union{AbstractVariableRef,GenericAffExpr}, rhs::Integer)
+    for (lhscoef, lhsvar) in linearterms(lhs)
+        for (rhscoef, rhsvar) in linearterms(rhs)
+            add_to_expression!(result, lhscoef*rhscoef, lhsvar, rhsvar)
         end
     end
 
     # Try to preallocate space for aff
     if !iszero(lhs.constant) && !iszero(rhs.constant)
-        sizehint!(ret.aff.vars, n+m)
-        sizehint!(ret.aff.coeffs, n+m)
+        sizehint!(result.aff, lhs_length + rhs_length)
     elseif !iszero(lhs.constant)
-        sizehint!(ret.aff.vars, n)
-        sizehint!(ret.aff.coeffs, n)
+        sizehint!(result.aff, rhs_length)
     elseif !iszero(rhs.constant)
-        sizehint!(ret.aff.vars, m)
-        sizehint!(ret.aff.coeffs, m)
+        sizehint!(result.aff, lhs_length)
     end
 
-    # [LHS constant] * RHS
+    # [LHS constant] * [RHS linear terms]
     if !iszero(lhs.constant)
         c = lhs.constant
-        for j = 1:m
-            push!(ret.aff.vars,   rhs.vars[j])
-            push!(ret.aff.coeffs, rhs.coeffs[j] * c)
+        for (rhscoef, rhsvar) in linearterms(rhs)
+            add_to_expression!(result.aff, c*rhscoef, rhsvar)
         end
-        ret.aff.constant += c * rhs.constant
     end
 
-    # [RHS constant] * LHS
+    # [RHS constant] * [LHS linear terms]
     if !iszero(rhs.constant)
         c = rhs.constant
-        for i = 1:n
-            push!(ret.aff.vars,   lhs.vars[i])
-            push!(ret.aff.coeffs, lhs.coeffs[i] * c)
+        for (lhscoef, lhsvar) in linearterms(lhs)
+            add_to_expression!(result.aff, c*lhscoef, lhsvar)
         end
-        # Don't do the following line
-        #ret.aff.constant += c * lhs.constant
-        # If lhs.constant is 0, its a waste of time
-        # If lhs.constant is non-zero, its already done
     end
 
-    return ret
+    result.aff.constant = lhs.constant * rhs.constant
+
+    return result
 end
 # GenericAffExpr--GenericQuadExpr
-Base.:+(a::GenericAffExpr, q::GenericQuadExpr) = GenericQuadExpr(copy(q.qvars1),copy(q.qvars2),copy(q.qcoeffs),a+q.aff)
-Base.:-(a::GenericAffExpr, q::GenericQuadExpr) = GenericQuadExpr(copy(q.qvars1),copy(q.qvars2),    -q.qcoeffs ,a-q.aff)
+Base.:+(a::GenericAffExpr, q::GenericQuadExpr) = GenericQuadExpr(a+q.aff, copy(q.terms))
+function Base.:-(a::GenericAffExpr, q::GenericQuadExpr)
+    result = -q
+    # This makes an unnecessary copy of aff, but it's important for a to appear
+    # first.
+    result.aff = a + result.aff
+    return result
+end
 
 # GenericQuadExpr
 Base.:+(lhs::GenericQuadExpr) = lhs
-Base.:-(lhs::GenericQuadExpr{T}) where T = zero(T)-lhs
+Base.:-(lhs::GenericQuadExpr) = map_coefficients(lhs, -)
 # GenericQuadExpr--Number
 Base.:+(lhs::GenericQuadExpr, rhs::Number) = (+)(+rhs,lhs)
 Base.:-(lhs::GenericQuadExpr, rhs::Number) = (+)(-rhs,lhs)
 Base.:*(lhs::GenericQuadExpr, rhs::Number) = (*)(rhs,lhs)
-Base.:/(lhs::GenericQuadExpr, rhs::Number) = (*)(inv(rhs),lhs)
+Base.:/(lhs::GenericQuadExpr, rhs::Number) = (*)(1.0/rhs,lhs)
```

Comment:
Why replacing `inv` ? Isn't it more specific ?

Reply:
Don't remember why I changed this.